From cebd9402de3cb0db6650ff8719dcda03533a613d Mon Sep 17 00:00:00 2001
From: Thu Nguyen <thu@os.amperecomputing.com>
Date: Wed, 2 Aug 2023 08:53:35 +0700
Subject: [PATCH 2/2] transport: Match type and command on response in
 pldm_transport_send_recv_msg()

mctp-demux broad cast the PLDM RX reponse messages to all of mctp
socket instances. The GetSensorReading response from one instance
(pldmd) can be received by the other socket instance (pldmtool) which
is waiting for the response of GetPDR. Only comparing the instance ID
is not enough because the instance ID of GetSensorReading and GetPDR
when the gap time between those PLDM commands is long enough.
Add the command type and command index on matching the response of
pldm_transport_send_recv_msg().

Signed-off-by: Thu Nguyen <thu@os.amperecomputing.com>
---
 src/transport/transport.c | 4 +++-
 1 file changed, 3 insertions(+), 1 deletion(-)

diff --git a/src/transport/transport.c b/src/transport/transport.c
index c552150..47fc5ce 100644
--- a/src/transport/transport.c
+++ b/src/transport/transport.c
@@ -191,7 +191,9 @@ pldm_transport_send_recv_msg(struct pldm_transport *transport, pldm_tid_t tid,
 					     resp_msg_len);
 		if (rc == PLDM_REQUESTER_SUCCESS) {
 			const struct pldm_msg_hdr *resp_hdr = *pldm_resp_msg;
-			if (req_hdr->instance_id == resp_hdr->instance_id) {
+			if (req_hdr->instance_id == resp_hdr->instance_id &&
+				req_hdr->type == resp_hdr->type &&
+				req_hdr->command == resp_hdr->command) {
 				return rc;
 			}
 
-- 
2.34.1

